prov/verbs/src/verbs_domain_xrc.c

```diff
@@ -0,0 +1,576 @@
+		if (!rsvd_qpn) {
+			VERBS_WARN(FI_LOG_FABRIC,
+				   "Create of XRC reserved QPN failed %d\n",
+				   errno);
+			return -errno;
+		}
+
+		memset(&open_attr, 0, sizeof(open_attr));
+		open_attr.qp_num = tgt_qpn;
+		open_attr.comp_mask = IBV_QP_OPEN_ATTR_NUM |
+			IBV_QP_OPEN_ATTR_XRCD | IBV_QP_OPEN_ATTR_TYPE |
+			IBV_QP_OPEN_ATTR_CONTEXT;
+		open_attr.xrcd = domain->xrc.xrcd;
+		open_attr.qp_type = IBV_QPT_XRC_RECV;
+		open_attr.qp_context = ep;
+
+		ep->tgt_ibv_qp = ibv_open_qp(domain->verbs, &open_attr);
+		if (!ep->tgt_ibv_qp) {
+			VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				   "XRC TGT QP, ibv_open_qp()", errno);
+			fi_ibv_release_qpn(rsvd_qpn);
+			return -errno;
+		}
+		ep->conn_setup->rsvd_tgt_qpn = rsvd_qpn;
+		return FI_SUCCESS;
+	}
+
+	/* An existing XRC target was not specified, create XRC TGT
+	 * side of new physical connection. */
+	fi_ibv_msg_ep_get_qp_attr(ep, NULL,
+			(struct ibv_qp_init_attr *)&attr_ex);
+	attr_ex.qp_type = IBV_QPT_XRC_RECV;
+	attr_ex.qp_context = ep;
+	attr_ex.comp_mask = IBV_QP_INIT_ATTR_PD | IBV_QP_INIT_ATTR_XRCD;
+	attr_ex.pd = domain->pd;
+	attr_ex.xrcd = domain->xrc.xrcd;
+	if (rdma_create_qp_ex(ep->tgt_id, &attr_ex)) {
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Physical XRC TGT QP, rdma_create_ep_ex()",
+				 errno);
+		return -errno;
+	}
+	ep->tgt_ibv_qp = ep->tgt_id->qp;
+
+	return FI_SUCCESS;
+}
+
+static int fi_ibv_put_tgt_qp(struct fi_ibv_ep *ep)
+{
+	int ret;
+
+	if (!ep->tgt_ibv_qp)
+		return FI_SUCCESS;
+
+	/* The kernel will not destroy the detached TGT QP until all
+	 * shared opens have called ibv_destroy_qp. */
+	ret = ibv_destroy_qp(ep->tgt_ibv_qp);
+	if (ret) {
+		VERBS_INFO_ERRNO(FI_LOG_EP_CTRL,
+				 "Close XRC TGT QP, ibv_destroy_qp()", errno);
+		return -errno;
+	}
+	ep->tgt_ibv_qp = NULL;
+
+	return FI_SUCCESS;
+}
+
+int fi_ibv_ep_destroy_xrc_qp(struct fi_ibv_ep *ep)
+{
+	struct fi_ibv_domain *domain = fi_ibv_msg_ep_to_domain(ep);
+
+	if (ep->ibv_qp) {
+		fastlock_acquire(&domain->xrc.ini_mgmt_lock);
+		fi_ibv_put_shared_ini_conn(ep);
+		fastlock_release(&domain->xrc.ini_mgmt_lock);
+	}
+	if (ep->id) {
+		rdma_destroy_id(ep->id);
+		ep->id = NULL;
+	}
+	if (ep->tgt_ibv_qp)
+		fi_ibv_put_tgt_qp(ep);
+
+	if (ep->tgt_id) {
+		rdma_destroy_id(ep->tgt_id);
+		ep->tgt_id = NULL;
+	}
+	return 0;
+}
+
+static int fi_ibv_ini_conn_compare(struct ofi_rbmap *map, void *key, void *data)
+{
+	struct fi_ibv_ini_shared_conn *ini_conn = data;
+	struct fi_ibv_ini_conn_key *_key = key;
+	int ret;
+
+	assert(_key->addr->sa_family == ini_conn->peer_addr->sa_family);
+
+	/* Only interested in the interface address and TX CQ */
+	switch (_key->addr->sa_family) {
+	case AF_INET:
+		ret = memcmp(&ofi_sin_addr(_key->addr),
+			     &ofi_sin_addr(ini_conn->peer_addr),
+			     sizeof(ofi_sin_addr(_key->addr)));
+		break;
+	case AF_INET6:
+		ret = memcmp(&ofi_sin6_addr(_key->addr),
+			     &ofi_sin6_addr(ini_conn->peer_addr),
+			     sizeof(ofi_sin6_addr(_key->addr)));
+		break;
```

Comment:
See ofi_net.h for helpers -- ofi_equals_ipaddr / ofi_equals_sockaddr

Reply:
Thanks, I've looked. If ok, I still prefer to use memcmp() directly:
1) I want to log the sa_family mis-match.
2) I want to return the actual memcmp result when not equal since ordering matters here.